PLFM-5812 (#591)
Sage-Bionetworks/synapseDocs: articles/using_synapse_as_an_oauth_server.md

---
title: "Using Synapse as an OAuth 2.0 Server"
layout: article
excerpt: Follow these steps to register an OAuth Client and authorize access to Synapse.
category: howto
---

<style>
#image {
    width: 100%;
}
#imageSmall {
    width: 40%;
}
</style>

# Registering and linking an OAuth 2.0 Client

> Synapse implements the [OpenID Connect](https://openid.net/specs/openid-connect-core-1_0.html) protocol, which is an extension of OAuth 2.0.  The Open ID Connect endpoint is `https://repo-prod.prod.sagebase.org/auth/v1`, and the Open ID Configuration (aka the "discovery document") is published in the standard location: [https://repo-prod.prod.sagebase.org/auth/v1/.well-known/openid-configuration](https://repo-prod.prod.sagebase.org/auth/v1/.well-known/openid-configuration).

Reviewer: These endpoints currently return a 'not found' error message - is that to be expected? When I hit https://repo-prod.prod.sagebase.org/auth/v1 I get:

```
{"reason":"GET /auth/v1 was not found. Please reference API documentation at https://docs.synapse.org/rest/"}
```

Author: Yes, it's to be expected.  The 'endpoint' is a prefix, not a URL.  I realize it's confusing so I am removing mention of this prefix.

> Synapse implements the [OpenID Connect](https://openid.net/specs/openid-connect-core-1_0.html) protocol, which is an extension of OAuth 2.0.  The Open ID Connect endpoint is `https://repo-prod.prod.sagebase.org/auth/v1`, and the Open ID Configuration (aka the "discovery document") is published in the standard location: [https://repo-prod.prod.sagebase.org/auth/v1/.well-known/openid-configuration](https://repo-prod.prod.sagebase.org/auth/v1/.well-known/openid-configuration).

Reviewer: Synapse does not implement this, the OpenID Connect protocol has been implemented in the Synapse platform.

Maybe this is a standard way to indicate this, but it just sounds funny to me! Ignore if you feel differently.

Author: OK

## Create an OAuth 2.0 Client

> You can create a client programmatically as follows:

Reviewer: ```suggestion
You can create a client as follows:
```

> ##### Python

Reviewer: Will there also be an `R` example?

Author: Ultimately there will be way to do it from the web browser, the preferred approach.  The Python instructions are just a placeholder until the web editor is in place, after which I'd like to remove the Python instructions from the article.  It's easy to rewrite the given instructions in R but I did not see a need for that.  Would you like to see (temporary) R instructions added here?

```python

> import synapseclient

Reviewer: Python code is not PEP-8 compliant or consistent, - snakecase, long lines, etc. Please lint! (except for existing Synapse client function/parameter names)

import json
syn = synapseclient.login()

clientMetaData = {
  'client_name': 'Your client name',
  'redirect_uris': [
    'https://yourhost.com/user/login'
  ],
  'client_uri': 'https://yourhost.com/index.html',
  'policy_uri': 'https://yourhost.com/policy',
  'tos_uri': 'https://yourhost.com/terms_of_service',
  'userinfo_signed_response_alg': 'RS256'
}

clientMetaData = syn.restPOST(uri='/oauth2/client', endpoint=syn.authEndpoint, body=json.dumps(clientMetaData))

client_id = clientMetaData['client_id']

clientIdAndSecret = syn.restPOST(uri='/oauth2/client/secret/'+client_id, endpoint=syn.authEndpoint, body='')

print(clientIdAndSecret)

```

The client URI, policy URI, and terms of service URI are optional, as is the `userinfo_signed_response_alg` parameter.  Further, you can optionally include a `sector_identifier_uri` parameter.  This is an advanced feature described [here](https://openid.net/specs/openid-connect-registration-1_0.html#SectorIdentifierValidation) and is relevant if the client uses multiple hosts because Synapse only returns `pairwise` subject values to its OAuth clients.

> The returned `client_id` and `client_secret` will be needed later when getting an access token.  The secret is only returned once.  (Synapse does not keep it.)  If lost, you can generate a new secret but the previous one will be invalidated.

Reviewer: ```suggestion
The returned `client_id` and `client_secret` will be needed later when getting an access token.  The secret is only returned once (It is not stored in Synapse). If lost, you can generate a new secret but the previous one will be invalidated.
```

Reviewer: Unclear how the previous one will be invalidated - is that if I create a new `clientMetaData` with the same information? What makes the `clientMetaData` unique - just the `name`, or combination of other attributes?

Author: > Unclear how the previous one will be invalidated - is that if I create a new clientMetaData with the same information?

No, not if you generate a new client, if you generate a new secret for the current client.

> What makes the clientMetaData unique - just the name, or combination of other attributes?

It's the `client_id` that uniquely identifies a client. Additionally, the client name cannot be reused.

You can retrieve, update, and delete your client programmatically as well:

```python
clientMetaData = syn.restGET(uri='/oauth2/client/'+client_id, endpoint=syn.authEndpoint)

clientMetaData['policy_uri'] = 'https://yourhost.com/updated_policy'

clientMetaData = syn.restPUT(uri='/oauth2/client/'+client_id, endpoint=syn.authEndpoint, clientMetaData)

> syn.restDELETE(uri='/oauth2/client/'+client_id, endpoint=syn.authEndpoint)

Reviewer: If this returns a value it should be captured for later inspection.

Author: it doesn't

```

To login via Synapse your client should redirect the browser from your application to `https://signin.synapse.org`, with the standard OAuth 2.0 request parameters:

- `client_id`=`<your client id>`
- `scope`=`openid`
- `redirect_uri`=`<the redirect uri registered with your client>`
- `response_type`=`code`
- `state`=`<any state you want returned>`
- `nonce`=`<some string to be returned in the ID token>`
- `claims`=`<a JSON object>`

> Synapse supports the `claims` request parameter, a JSON document containing the details of the user identity information you would like returned, as described [here](https://openid.net/specs/openid-connect-core-1_0.html#ClaimsParameter).  The list of supported claims is given here TBD.  For most claims the value to include in the JSON document is `null`.  The exception is the `team` claim, for which you provide the IDs of one or more teams, the membership of which you wish to inquire about.  Synapse will return the IDs of the subset of the given list of teams to which the user belongs.  Here is an example of a claims parameter JSON object:

Reviewer: ```suggestion
Synapse supports the `claims` request parameter, a JSON document containing the details of the user identity information you would like returned, as described [here](https://openid.net/specs/openid-connect-core-1_0.html#ClaimsParameter). The list of supported claims is to be determined. For most claims the value to include in the JSON document is `null`. The exception is the `team` claim, for which you provide the IDs of one or more teams, the membership of which you wish to inquire about. Synapse will return the IDs of the subset of the given list of teams to which the user belongs. Here is an example of a claims parameter JSON object:
```

The statement that the list of supported claims is 'to be determined' should be qualified by an open issue either in this repository or to a Jira issue.

Author: @kdaily I put "TBD" simply because I was waiting for the API doc's to be published.  I have now filled the placeholder in with the required URL.

```
 {
   "id_token":
    {
     "given_name": null,
     "family_name": null,
     "is_certified": null,
     "team": {"values":["273957", "3385814"]}
    },
   "userinfo":
    {
     "given_name": null,
     "family_name": null,
     "is_certified": null,
     "team": {"values":["273957", "3385814"]}
    }
  }
```

> Following the user's consent, Synapse will redirect back to your specified redirect URI with an authorization code and your 'state' data.  You can then exchange the authorization code for a ID token and access token in the standard way.  Here is an example using cUrl:

Reviewer: I don't understand the word `consent` here.

Author: The user has to click "Allow" in the web browser for Synapse to send the info to the linked client application.

> Following the user's consent, Synapse will redirect back to your specified redirect URI with an authorization code and your 'state' data.  You can then exchange the authorization code for a ID token and access token in the standard way.  Here is an example using cUrl:

Reviewer: ```suggestion
Following the user's consent, Synapse will redirect back to your specified redirect URI with an authorization code and your 'state' data.  You can then exchange the authorization code for a ID token and access token in the standard way.  Here is an example using `curl`:
```
